src/AzureRenderManager/Microsoft.Identity.Web/StartupHelpers.cs

```diff
@@ -0,0 +1,228 @@
+                // (on which the original .NET core template is based) are different than Azure AD V2 endpoint.
+                // For more details see [ID Tokens](https://docs.microsoft.com/en-us/azure/active-directory/develop/id-tokens)
+                // and [Access Tokens](https://docs.microsoft.com/en-us/azure/active-directory/develop/access-tokens)
+                options.TokenValidationParameters.NameClaimType = "preferred_username";
+
+                // Force the account selection (to avoid automatic sign-in with the account signed-in with Windows)
+                //options.Prompt = "select_account";
+
+                // Avoids having users being presented the select account dialog when they are already signed-in
+                // for instance when going through incremental consent
+                options.Events.OnRedirectToIdentityProvider = context =>
+                {
+                    var login = context.Properties.GetParameter<string>(OpenIdConnectParameterNames.LoginHint);
+                    if (!string.IsNullOrWhiteSpace(login))
+                    {
+                        context.ProtocolMessage.LoginHint = login;
+                        context.ProtocolMessage.DomainHint = context.Properties.GetParameter<string>(OpenIdConnectParameterNames.DomainHint);
+
+                        // delete the login_hint and domainHint from the Properties when we are done otherwise
+                        // it will take up extra space in the cookie.
+                        context.Properties.Parameters.Remove(OpenIdConnectParameterNames.LoginHint);
+                        context.Properties.Parameters.Remove(OpenIdConnectParameterNames.DomainHint);
+                    }
+
+                    // Additional claims
+                    if (context.Properties.Items.ContainsKey(OidcConstants.AdditionalClaims))
+                    {
+                        context.ProtocolMessage.SetParameter(OidcConstants.AdditionalClaims,
+                                                             context.Properties.Items[OidcConstants.AdditionalClaims]);
+                    }
+
+                    return Task.FromResult(0);
+                };
+            });
+            return services;
+        }
+
+        // Because we don't have a persistent token cache, after a web app
+        // restart the user can have a valid token set in a cookie, but the 
+        // server side TokenCache doesn't have the refresh token and hence we cannot
+        // get Graph or ARM resource tokens.
+        // For this reason we set a marker to indicate the user has logged into 
+        // this instance of the web app at least once.
+        private static Task SetIncarnationCacheKey(CookieSigningInContext context)
+        {
+            var cache = context.HttpContext.RequestServices.GetRequiredService<IMemoryCache>();
+            var cacheKey = CacheKey(GetObjectId(context.Principal.Claims));
```

Comment:
Some claims don't have object ids (strange multi-tenant guest scenarios). Do we need to check for that at all, or does the rest of the token checks / initial login require an ObjectId?

Reply:
Fixed with:             
string accountIdentifier = claimsPrincipal.GetMsalAccountId();
            if (accountIdentifier == null)
            {
                accountIdentifier = claimsPrincipal.GetLoginHint();
            }